src/KilnUniV2LPSwap.sol

```diff
@@ -0,0 +1,133 @@
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+pragma solidity ^0.8.14;
+
+import {KilnBase, GemLike} from "./KilnBase.sol";
+import {UniswapV2Library} from "./uniV2/UniswapV2Library.sol";
+
+// Resources:
+//    https://docs.uniswap.org/contracts/v2/reference/smart-contracts/library
+//    https://docs.uniswap.org/contracts/v2/reference/smart-contracts/router-02
+
+interface UniswapRouterV2Like {
+    function factory() external returns (address);
+    function swapExactTokensForTokens(
+        uint amountIn,
+        uint amountOutMin,
+        address[] calldata path,
+        address to,
+        uint deadline
+    ) external returns (uint[] memory amounts);
+    function addLiquidity(
+        address tokenA,
+        address tokenB,
+        uint amountADesired,
+        uint amountBDesired,
+        uint amountAMin,
+        uint amountBMin,
+        address to,
+        uint deadline
+    ) external returns (uint amountA, uint amountB, uint liquidity);
+    function removeLiquidity(
+        address tokenA,
+        address tokenB,
+        uint liquidity,
+        uint amountAMin,
+        uint amountBMin,
+        address to,
+        uint deadline
+    ) external returns (uint amountA, uint amountB);
+}
+
+contract KilnUniV2LPSwap is KilnBase {
+
+    using UniswapV2Library for *;
+
+    address public immutable uniV2Router;
+    address public immutable pairToken;
+    address public immutable receiver;
+
+    uint256 internal constant WAD = 10 ** 18;
+
+    uint256 public max;  // [WAD] Maximum acceptable buy price in sell (WAD)
+
+    constructor(address _sell, address _buy, address _uniV2Router, address _receiver) KilnBase(_sell, _buy) {
+        receiver = _receiver;
+        uniV2Router = _uniV2Router;
+        pairToken = UniswapRouterV2Like(_uniV2Router).factory().pairFor(_sell, _buy);
+    }
+
+    //rounds to zero if x*y < WAD / 2
+    function _wmul(uint x, uint y) internal pure returns (uint z) {
+        z = (x * y) + (WAD / 2) / WAD;
+    }
+
+    function file(bytes32 what, uint256 data) public override auth {
+        if (what == "max") {
+            max = data;
+        } else {
+            super.file(what, data);
+            return;
+        }
+        emit File(what, data);
+    }
+
+    function _swap(uint256 _amount) internal override returns (uint256 _swapped) {
+
+        uint256 _halfLot = _amount / 2;
+
+        GemLike(sell).approve(uniV2Router, _amount);
+
+        address[] memory _path = new address[](2);
+        _path[0] = sell;
+        _path[1] = buy;
+
+        uint256 _amountOutMin = _wmul(_halfLot, max);
```

Comment:
if max is a price denominated in WAD then we need to divide by WAD here as well.